src/inc/til/unicode.h

```diff
@@ -0,0 +1,154 @@
+        return (wch & 0xFC00) == 0xD800;
+    }
+
+    static constexpr bool is_trailing_surrogate(const wchar_t wch) noexcept
+    {
+        return (wch & 0xFC00) == 0xDC00;
+    }
+
+    constexpr std::wstring_view utf16_next(std::wstring_view wstr) noexcept
+    {
+        auto it = wstr.begin();
+        const auto end = wstr.end();
+        auto p = L"\uFFFD";
+        size_t l = 1;
+
+        if (it != end)
+        {
+            const auto wch = *it;
+            p = &*it;
+
+            if (is_surrogate(wch))
+            {
+                ++it;
+                const auto wch2 = it != end ? *it : wchar_t{};
+                if (is_leading_surrogate(wch) && is_trailing_surrogate(wch2))
+                {
+                    l = 2;
+                    ++it;
+                }
+                else
+                {
+                    p = L"\uFFFD";
+                }
+            }
+        }
+
+        return { p, l };
+    }
+
+    struct utf16_iterator
+    {
+        struct sentinel
+        {
+        };
+
+        struct iterator
+        {
+            using iterator_category = std::forward_iterator_tag;
+            using value_type = std::wstring_view;
+            using reference = value_type&;
+            using pointer = value_type*;
+            using difference_type = std::ptrdiff_t;
+
+            explicit constexpr iterator(utf16_iterator& p) noexcept :
+                _p{ p }
+            {
+            }
+
+            const value_type& operator*() const noexcept
+            {
+                return _p.value();
+            }
+
+            iterator& operator++() noexcept
+            {
+                _p._advance = true;
+                return *this;
+            }
+
+            bool operator!=(const sentinel&) const noexcept
+            {
+                return _p.valid();
+            }
+
+        private:
+            utf16_iterator& _p;
+        };
+
+        explicit constexpr utf16_iterator(std::wstring_view wstr) noexcept :
+            _it{ wstr.begin() }, _end{ wstr.end() }, _advance{ _it != _end }
+        {
+        }
+
+        iterator begin() noexcept
+        {
+            return iterator{ *this };
+        }
+
+        sentinel end() noexcept
+        {
+            return sentinel{};
+        }
+
+    private:
+        bool valid() const noexcept
+        {
+            return _it != _end;
+        }
+
+        void advance() noexcept
+        {
+            const auto wch = *_it;
+            auto p = &*_it;
+            size_t l = 1;
+
+            ++_it;
+
+            if (is_surrogate(wch))
+            {
+                const auto wch2 = _it != _end ? *_it : wchar_t{};
+                if (is_leading_surrogate(wch) && is_trailing_surrogate(wch2))
+                {
+                    l = 2;
+                    ++_it;
+                }
+                else
+                {
+                    p = L"\uFFFD";
+                }
+            }
+
+            _v = { p, l };
```

Comment:
I mean, I get that `advance` isn't const, but isn't it weird that we're replacing the data with the replacement char? Or is it intentional that this is basically 1-to-1 the same as `utf16_next()`?

Reply:
Yeah this is intentional. `_v` is basically the value that the iterator will return in the next iteration and if the unicode data is invalid it will contain U+FFFD for just one iteration. In the next one after, it'll (hopefully) return some valid characters again.